components/camel-univocity-parsers/src/main/java/org/apache/camel/dataformat/univocity/Marshaller.java

```diff
@@ -86,7 +88,9 @@ private void writeRow(Exchange exchange, Object row, W writer) throws NoTypeConv
         Map<?, ?> map = convertToMandatoryType(exchange, Map.class, row);
         if (adaptHeaders) {
             for (Object key : map.keySet()) {
-                headers.add(convertToMandatoryType(exchange, String.class, key));

```

Comment:
Seems like the marshaller has to keep state between rows because it must be able to add headers while keeping the original order - so says the comment on that unit test: "Tests that the marshalling adds new columns on the fly and keep its order".